module/state_synchronization/requester/oneshot_execution_data_requester.go

```diff
@@ -0,0 +1,182 @@
+	// the initial delay used in the exponential backoff for failed execution data download
+	// retries.
+	RetryDelay time.Duration
+	// the max delay used in the exponential backoff for failed execution data download.
+	MaxRetryDelay time.Duration
+}
+
+// OneshotExecutionDataRequester is a component that requests execution data for a block.
+// It uses a retry mechanism to retry the download execution data if they are not found.
+type OneshotExecutionDataRequester struct {
+	log             zerolog.Logger
+	metrics         module.ExecutionDataRequesterMetrics
+	config          OneshotExecutionDataConfig
+	execDataCache   *cache.ExecutionDataCache
+	executionResult *flow.ExecutionResult
+	blockHeader     *flow.Header
+}
+
+func NewOneshotExecutionDataRequester(
+	log zerolog.Logger,
+	metrics module.ExecutionDataRequesterMetrics,
+	execDataCache *cache.ExecutionDataCache,
+	executionResult *flow.ExecutionResult,
+	blockHeader *flow.Header,
+	config OneshotExecutionDataConfig,
+) (*OneshotExecutionDataRequester, error) {
+	if blockHeader.ID() != executionResult.BlockID {
+		return nil, fmt.Errorf("block id and execution result mismatch")
+	}
+
+	return &OneshotExecutionDataRequester{
+		log:             log.With().Str("component", "oneshot_execution_data_requester").Logger(),
+		metrics:         metrics,
+		execDataCache:   execDataCache,
+		executionResult: executionResult,
+		blockHeader:     blockHeader,
+		config:          config,
+	}, nil
+}
+
+// RequestExecutionData requests execution data for a given block from the execution data cache.
+// It performs a fetch using a retry mechanism with exponential backoff if execution data not found.
+// Returns the execution data entity and any error encountered.
+//
+// The function logs each retry attempt and emits metrics for retries and fetch durations.
+// The block height is used only for logging and metric purposes.
+//
+// Expected errors:
+// - context.Canceled: if the provided context was canceled before completion
+// All other errors are unexpected exceptions and may indicate invalid execution data was received.
+func (r *OneshotExecutionDataRequester) RequestExecutionData(
+	ctx context.Context,
+) (*execution_data.BlockExecutionDataEntity, error) {
+	backoff := retry.NewExponential(r.config.RetryDelay)
+	backoff = retry.WithCappedDuration(r.config.MaxRetryDelay, backoff)
+	backoff = retry.WithJitterPercent(15, backoff)
+
+	// bitswap always waits for either all data to be received or a timeout, even if it encountered an error.
+	// use an incremental backoff for the timeout so we do faster initial retries, then allow for more
+	// time in case data is large or there is network congestion.
+	timeout := retry.NewExponential(r.config.FetchTimeout)
+	timeout = retry.WithCappedDuration(r.config.MaxFetchTimeout, timeout)
+
+	attempt := 0
+	blockID := r.executionResult.BlockID
+	blockHeight := r.blockHeader.Height
+	executionDataID := r.executionResult.ExecutionDataID
```

Comment:
nit: since there are more than 1 log message that include these,  may as well create a local logger
```suggestion
	lg := r.log.With().
		Str("block_id", r.executionResult.BlockID.String()).
		Str("execution_data_id", r.executionResult.ExecutionDataID.String()).
		Uint64("height", r.blockHeader.Height).
		Logger()
	
```